### PR TITLE
[WIP] LPS-88530 Make endpoint configurable to iOS client 

### PR DESCRIPTION
**Pull request by mclomello**

### CODE REVIEW

File: Sample/Podfile.lock [hunk 1]
```diff
@@ -1,11 +1,11 @@
 PODS:
-  - liferay-analytics-forms-ios (0.1.1):
-    - liferay-analytics-ios (= 0.1.0)
+  - liferay-analytics-forms-ios (1.0.0):
```
Comment (victorlaerte) [inline Sample/Podfile.lock:2]: Is there any change in the plugin to bump the version? I thought it was other story

File: Source/Api/Analytics.swift [hunk 2]
```diff
@@ -72,8 +102,7 @@ public class Analytics {
 	*/
 	public class func setIdentity(email: String, name: String = "") {
 		let instance = try! Analytics.getInstance()
-		
-		let identityContext = instance.getDefaultIdentityContext()
+		let identityContext = IdentityContext(dataSourceId: instance.dataSourceId)
```
Comment (victorlaerte) [inline Source/Api/Analytics.swift]: Where is the `platform` and `language` initialization that the old method does?

File: Source/Api/Analytics.swift [hunk 3]
```diff
@@ -107,11 +138,11 @@ public class Analytics {
 	}
 	
 	class func getInstance() throws -> Analytics {
-		guard let sharedInstance = Analytics.sharedInstance else {
+		guard let instance = Analytics.instance else {
```
Comment (victorlaerte) [inline Source/Api/Analytics.swift]: `sharedInstance` is how Swift convention call Singletons, you should not change this.

File: Source/Model/AnalyticsEvents.swift [hunk 4]
```diff
@@ -17,16 +17,16 @@
...
     
-	init(analyticsKey: String, userId: String, build: (AnalyticsEvents) -> Void) {
-        self.analyticsKey = analyticsKey
+	init(dataSourceId: String, userId: String, build: (AnalyticsEvents) -> Void) {
+		self.dataSourceId = dataSourceId
 		self.userId = userId
 
-        build(self)
-    }
+		build(self)
```
Comment (victorlaerte) [inline Source/Model/AnalyticsEvents.swift:24]: Are you using tabs in `Preferences > Text Editing > Prefer indent using`?

Please reverse indentation in this file.

File: Source/Model/Event.swift [hunk 5]
```diff
@@ -21,12 +21,13 @@ internal class Event: Codable {
 	init(applicationId: String, eventId: String, build: ((Event) -> Void)? = nil) {
 		self.applicationId = applicationId
 		self.eventId = eventId
-
+		self.eventDate = DateUtil.formatDate(date: self.eventDate)
```
Comment (victorlaerte) [inline Source/Model/Event.swift]: Why this?

File: Source/Model/Event.swift [hunk 6]
```diff
@@ -21,12 +21,13 @@ internal class Event: Codable {
...
+		
 		build?(self)
     }
-    
+
     let applicationId: String
-    let eventDate = Date()
+    var eventDate = Date()
```
Comment (victorlaerte) [inline Source/Model/Event.swift]: isn't possible to keep?
`let eventDate = DateUtil.formatDate(Date())`

File: Source/Resources/Info.plist [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (victorlaerte) [inline Source/Resources/Info.plist]: Do we need to add this file?

File: Source/Util/DateUtil.swift [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+* details.
+*/
+
+import Foundation
+
+/**
+- Author: Marcelo Mello
+*/
+internal class DateUtil {
```
Comment (victorlaerte) [inline Source/Util/DateUtil.swift]: I think it's better to create extension classes instead of utility classes.

File: Source/Util/DateUtil.swift [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	static let DATE_FORMAT: String = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
+	static let DATE_ABBREVIATION: String = "UTC"
+	
+	static func formatDate(date: Date) -> Date {
```
Comment (victorlaerte) [inline Source/Util/DateUtil.swift]: This maybe can be an extension function called `formatAsUTC`

File: Source/Util/FlushProcess.swift [hunk 10]
```diff
@@ -99,8 +100,9 @@ internal class FlushProcess {
 			isInProgress = true
 			
 			var userIdsEvents = eventsDAO.getEvents()
+			
 			for (userId, currentEvents) in userIdsEvents {
-				try send(userId: userId, events: currentEvents)
+				try sendEventsForUserId(userId: userId, events: currentEvents)
```
Comment (victorlaerte) [inline Source/Util/FlushProcess.swift]: If you use the parameters in the method name, named parameters are useless. You are just repeating the same information.

File: Source/Util/FlushProcess.swift [hunk 11]
```diff
@@ -74,20 +75,20 @@ internal class FlushProcess {
 		}
 	}
 	
-	func send(userId: String, events: [Event]) throws {
+	func sendEventsForUserId(userId: String, events: [Event]) throws {
```
Comment (victorlaerte) [inline Source/Util/FlushProcess.swift]: Swift supports idiomatic method signatures, for example:

`func send(events: [Event], for userId: String)`

File: Source/Util/URLUtil.swift [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import Foundation
+
+/**
+ - Author: Marcelo Mello
+ */
+internal class URLUtil {
```
Comment (victorlaerte) [inline Source/Util/URLUtil.swift]: URL class extension would be better

File: Source/Util/URLUtil.swift [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    
+    static func isValidUrl(value: String) -> Bool {
+        if value.hasPrefix("http://") || value.hasPrefix("https://") {
```
Comment (victorlaerte) [inline Source/Util/URLUtil.swift]: This method is doesn't mean it's a valid URL.

from `URL` constructor:

```
public struct URL : ReferenceConvertible, Equatable {

    /// Initialize with string.
    ///
    /// Returns `nil` if a `URL` cannot be formed with the string (for example, if the string contains characters that are illegal in a URL, or is an empty string).
    public init?(string: String)
```

**Review by victorlaerte — CHANGES_REQUESTED** [review]

File: Sample/Podfile.lock (hunk 1 above)
Comment (mclomello) [reply Sample/Podfile.lock:2]: That was a automatic refresh when I changed the folder structure for resources and ran "pod install" command. Can I set the version manually to 0.1.2? Seems more correct to me.

File: Source/Api/Analytics.swift (hunk 2 above)
Comment (mclomello) [reply Source/Api/Analytics.swift]: I changed that just for simplify the tests when I starting the changes, but the old method "getDefaultIdentityContext()" still there and I'll put back until finish the changes.

File: Source/Api/Analytics.swift (hunk 3 above)
Comment (mclomello) [reply Source/Api/Analytics.swift]: Oh, sorry. I don't knew that convention. I'll reverse this change.

File: Source/Model/AnalyticsEvents.swift (hunk 4 above)
Comment (mclomello) [reply Source/Model/AnalyticsEvents.swift:24]: In fact, when I modify this file first, my Xcode indentation was wrong, but I already fix that.
I made this change to correct this in SF commit. 
If I reverse that change, the indentation will be wrong.

File: Source/Model/Event.swift (hunk 5 above)
Comment (mclomello) [reply Source/Model/Event.swift]: I thought when create a Event instance, we can put eventDate to the expected format for Analytics. That can be a problem?

File: Source/Util/DateUtil.swift (hunk 9 above)
Comment (mclomello) [reply Source/Util/DateUtil.swift]: Yeah, I'll research and make this change.

File: Source/Util/FlushProcess.swift (hunk 10 above)
Comment (mclomello) [reply Source/Util/FlushProcess.swift]: This change in particular was just to keep the same name and pattern as the android library. 
I can reverse that.

File: Source/Util/FlushProcess.swift (hunk 11 above)
Comment (mclomello) [reply Source/Util/FlushProcess.swift]: Nice! I didn't knew that. I'll research about this and modify the method signatures that makes sense. Thanks!

File: Source/Util/URLUtil.swift (hunk 12 above)
Comment (mclomello) [reply Source/Util/URLUtil.swift]: Yeah, I'll research and make this change.

File: Source/Util/URLUtil.swift (hunk 13 above)
Comment (mclomello) [reply Source/Util/URLUtil.swift]: Yeah, that's a partial implementation. I'm working to fix that until finish the changes.

File: Sample/Podfile.lock (hunk 1 above)
Comment (victorlaerte) [reply Sample/Podfile.lock:2]: No, I think it's ok. Pod got the latest version

File: Source/Api/Analytics.swift (hunk 2 above)
Comment (victorlaerte) [reply Source/Api/Analytics.swift]: The method still there but in this place, you are using the constructor, so platform and language will not be initialized properly

File: Source/Model/Event.swift (hunk 5 above)
Comment (victorlaerte) [reply Source/Model/Event.swift]: I don't think so, but is it possible to keep it let as I've mentioned in the next comment?

Comment (mclomello) [reply Source/Model/Event.swift]: Yeah, for sure!

File: Source/Util/DateUtil.swift (hunk 8 above)
Comment (mclomello) [reply Source/Util/DateUtil.swift]: Yes. For sure!

File: Source/Model/Event.swift (hunk 6 above)
Comment (mclomello) [reply Source/Model/Event.swift]: Yes, we can.
Will be like the code below
`let eventDate = Date.formatAsUTC(date: Date())`

File: Source/Resources/Info.plist (hunk 7 above)
Comment (mclomello) [reply Source/Resources/Info.plist]: I think we need for override the properties. I'll research more and look if we need that file at all.
This file previously was named "settings.plist".

File: Source/Api/Analytics.swift (hunk 2 above)
Comment (mclomello) [reply Source/Api/Analytics.swift]: Yes. I'll back to get the defaultIdentityContext.

File: Sample/analytics-ios-demo/Assets.xcassets/AppIcon.appiconset/Contents.json [hunk 14]
```diff
@@ -1,93 +1,98 @@
...
+    {
+      "idiom" : "ios-marketing",
+      "size" : "1024x1024",
+      "scale" : "1x"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
```
Comment (phcp) [inline Sample/analytics-ios-demo/Assets.xcassets/AppIcon.appiconset/Contents.json]: SF problem in this file.

File: Source/Api/Impl/AnalyticsClient.swift [hunk 15]
```diff
@@ -18,31 +18,20 @@ import Foundation
...
 		encoder.dateEncodingStrategy = .iso8601
 		let analyticsData = try encoder.encode(analyticsEvents)
-			
-		guard let url = baseUrl else {
-			throw HttpError.invalidUrl
-		}
 		
 		let (data, _, error) = URLSession.sendPost(url: url, body: analyticsData)
-
+
```
Comment (phcp) [inline Source/Api/Impl/AnalyticsClient.swift]: There are some lines just with a tab added when it's not needed

File: Source/Api/Analytics.swift [hunk 16]
```diff
@@ -46,21 +47,50 @@ public class Analytics {
...
-	public class func configure(
-		analyticsKey: String, flushInterval: Int = Analytics.DEFAUL_TIME_INTERVAL) throws {
-		
+	public class func `init`(flushInterval: Int = Analytics.DEFAUL_TIME_INTERVAL) throws {
+        
 		if let _ = sharedInstance {
 			throw AnalyticsError.analyticsAlreadyInitialized
 		}
-		
+
```
Comment (phcp) [inline Source/Api/Analytics.swift]: The same tab problem here.

File: Source/Api/Analytics.swift [hunk 17]
```diff
@@ -46,21 +47,50 @@ public class Analytics {
...
+		}
+
+		guard let dataSourceId = settings?["com.liferay.analytics.DataSourceId"] else {
+			throw AnalyticsError.dataSourceIdNullOrEmpty
+		}
+
+		var endpointURL: String = ""
+
+		if let stringURL = settings?["com.liferay.analytics.EndpointUrl"] {
+			guard let _ = URL(string: stringURL) else {
```
Comment (phcp) [inline Source/Api/Analytics.swift]: Would be better to unify the if let with the guard let.

File: Source/Api/Impl/IdentityClient.swift [hunk 18]
```diff
@@ -19,22 +19,13 @@ import Foundation
...
-		baseUrl = URL(string: settings?["IDENTITY_GATEWAY"] ?? "")
-	}
-	
-	func send(identityContext: IdentityContext) throws {
-		guard let url = baseUrl else {
+		guard let url = URL(string: endpointURL) else {
 			throw HttpError.invalidUrl
 		}
-			
+
```
Comment (phcp) [inline Source/Api/Impl/IdentityClient.swift]: Tab problem here.

File: Source/Util/FlushProcess.swift [hunk 19]
```diff
@@ -74,20 +74,21 @@ internal class FlushProcess {
 		}
 	}
 	
-	func send(userId: String, events: [Event]) throws {
+	func send(events: [Event], for userId: String) throws {
```
Comment (phcp) [inline Source/Util/FlushProcess.swift:77]: 👏👏👏

File: Source/Util/FlushProcess.swift [hunk 20]
```diff
@@ -74,20 +74,21 @@ internal class FlushProcess {
...
 		let instance = try! Analytics.getInstance()
 		
 		var currentEvents = events
 		while !currentEvents.isEmpty {
 			let eventsToSend = Array(currentEvents.prefix(FLUSH_SIZE))
 			
 			let analyticsEvents = AnalyticsEvents(
-			analyticsKey: instance.analyticsKey, userId: userId) {
+			dataSourceId: instance.dataSourceId, userId: userId) {
```
Comment (phcp) [inline Source/Util/FlushProcess.swift]: SF

File: Tests/Impl/AnalyticsClientTest.swift [hunk 21]
```diff
@@ -13,33 +13,40 @@
...
-				$0.protocolVersion = "1.0"
-		}
-		
 		do {
-			let _ = try _analyticsClient.send(analyticsEvents: analyticsEvents)
+			try Analytics.init()
+			let instance = try Analytics.getInstance()
+			
+			let body = ["status": "success"]
+			stub(http(.post, uri: instance.endpointURL), json(body))
```
Comment (phcp) [inline Tests/Impl/AnalyticsClientTest.swift:34]: I liked that 👏

**Review by phcp — CHANGES_REQUESTED** [review]
Just some small changes. Thanks @mclomello

File: Sample/analytics-ios-demo/Assets.xcassets/AppIcon.appiconset/Contents.json (hunk 14 above)
Comment (mclomello) [reply Sample/analytics-ios-demo/Assets.xcassets/AppIcon.appiconset/Contents.json]: You're right, I'll fix this, thanks.

File: Source/Api/Impl/AnalyticsClient.swift (hunk 15 above)
Comment (mclomello) [reply Source/Api/Impl/AnalyticsClient.swift]: I'll fix that. Probably happens when my Xcode was not configured. Thanks

File: Source/Api/Analytics.swift (hunk 17 above)
Comment (mclomello) [reply Source/Api/Analytics.swift]: Yeah Paulo. I didn't knew that can be done. Thanks for warning and help me. I already the guard let.

**Pull request merged**